presto-spi/src/main/java/com/facebook/presto/spi/plan/OrderingScheme.java

```diff
@@ -11,76 +11,75 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package com.facebook.presto.sql.planner;
+package com.facebook.presto.spi.plan;
 
 import com.facebook.presto.spi.block.SortOrder;
 import com.facebook.presto.spi.relation.VariableReferenceExpression;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
-import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.ImmutableMap;
-import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Maps;
 
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
 
-import static com.google.common.base.MoreObjects.toStringHelper;
-import static com.google.common.base.Preconditions.checkArgument;
-import static com.google.common.collect.ImmutableList.toImmutableList;
-import static com.google.common.collect.Iterables.getOnlyElement;
 import static java.lang.String.format;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Collections.unmodifiableMap;
+import static java.util.Collections.unmodifiableSet;
 import static java.util.Objects.requireNonNull;
 
 public class OrderingScheme
 {
     private final List<VariableReferenceExpression> orderBy;
     private final Map<VariableReferenceExpression, SortOrder> orderings;
 
+    /**
+     * OrderingScheme indicates the sort order of rows:
+     * `orderby`:  keys to sort
+     * `orderings`: sort directions of keys
+     */
     @JsonCreator
-    public OrderingScheme(@JsonProperty("orderBy") List<VariableReferenceExpression> orderBy, @JsonProperty("orderings") Map<VariableReferenceExpression, SortOrder> orderings)
+    public OrderingScheme(
+            @JsonProperty("orderBy") List<VariableReferenceExpression> orderBy,
```

Comment:
I prefer `OrderingScheme to take a list of (`VariableReferenceExpression`, `SortOrder`) pairs, and create the list and map in constructor. That way there's no room for mistake and the `checkArgument` on keySet will not be necessary.

Reply:
Since we don't use javafx, I see we should make our own Pair class, which other classes also do with `Pair` being an inner class. However, since this class will be used to construct pairs BEFORE we construct OrderingScheme, this means other classes would have `... new OrderingScheme.Pair<K,V>...` which is messy IMO. Do we leave this or would it be reasonable to make a new class for Pair in SPI.

Maybe we could use `LinkedHashMap` for the interface. James mentioned I should verify if  json serde can create a `LinkedHashMap` if we do this (if not, create a serializer/deserializer).

What do you believe would be the best option?

Reply:
A `VariableReferenceExpression`, `SortOrder` pair is one "Ordering" in the `OrderingScheme` so it doesn't need to be named `Pair`. It can be a class with proper name. Using `LinkedHashMap` is quite implicit. People might or might not realize it's ordered immediately.